### PR TITLE
TDD problem. Finish problem and add additional tests for optional attendees.

### PR DESCRIPTION
**Pull request by sasymer**
- Finish problem using a bit cleaner solution than before
- Add tests for optional attendees
- Revise code to consider optional attendees

### CODE REVIEW

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 1]
```diff
@@ -20,118 +6,45 @@
...
+    } else {
+      return getReturnTimes(attendees, events, duration);
     }
+  }
 
+  private Collection<TimeRange> getReturnTimes(Collection<String> attendees, Collection<Event> events, long duration) {
     Collection<TimeRange> returnTimes = new ArrayList<>();
-
     List<Event> eventsList = new ArrayList<>(events);
     int eventIndex = 0;
```
Comment (wdcroslen) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:59]: I think the Indexes as well as the upper and lower bounds are clever to go through the event as well as add the Time Ranges in your list very clearly!

File: walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java [hunk 2]
```diff
@@ -319,46 +373,19 @@ public void optionalWithGaps() {
...
+      new Event("Event 3", TimeRange.fromStartEnd(TIME_0900AM, TimeRange.END_OF_DAY, true),
+          Arrays.asList(PERSON_C)));
+    
+    MeetingRequest request = new MeetingRequest(Arrays.asList(), DURATION_30_MINUTES);
+    request.addOptionalAttendee(PERSON_A);
+    request.addOptionalAttendee(PERSON_B);
+    request.addOptionalAttendee(PERSON_C);
 
     Collection<TimeRange> actual = query.query(events, request);
     Collection<TimeRange> expected = Arrays.asList();
```
Comment (wdcroslen) [inline walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java:391]: Overall good changes!

**Review by wdcroslen — APPROVED** [review]

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 3]
```diff
@@ -20,118 +6,45 @@
 import java.util.List;
 import java.util.Set;
 
+/*
+  Assumptions: the collection of events is in sorted order by start time
```
Comment (gagomez) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Let's call these preconditions and put them on the methods that require this state.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 4]
```diff
@@ -20,118 +6,45 @@
...
-      } else {
-        condensedBlockedTimes.add(first);
-      }
-    }
-    return condensedBlockedTimes;
-  }
-
-  public Collection<TimeRange> otherway(Collection<Event> events, MeetingRequest request) {
+  public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
     Collection<String> attendees = request.getAttendees();
```
Comment (gagomez) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: nit: it's hard to differentiate attendees from allAttendees, I would name this one "requiredAttendees"

**Review by gagomez — APPROVED** [review]

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 3 above)
Comment (sasymer) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Done

**Review by jiaxinz-g — APPROVED** [review]

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 4 above)
Comment (sasymer) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Named it mandatoryAttendees - done

**Pull request merged**